translate-c: missing semicolon

The issue here is that the translated C header file contains a Zig declaration that is missing a semicolon for some reason.  This example was minified from trying to include `windows.h`.  I tried to minify it as much as I could but this is as small as I could get it.  Note that this specific error only seems to occur on Windows (on linux I get other compile errors).

bug.h
```c
__forceinline struct SomeStruct *Func1(void)
{
    return (void*)__readgsqword(0);
}
void Func2(void)
{
    __readgsqword(0);
}
```
bug.zig
```zig
const c = @cImport({
    @cInclude("bug.h");
});

pub fn main() void {
    c.Func2();
}
```

```batch
> zig build-exe bug.zig -lc -I.
zig-cache\o\93806b0464444cfefd5c8574d1f69fa7\cimport.zig:3:1: error: expected token ';', found 'pub'
pub const Func1 = @compileError("unable to translate function"); // .\bug.h:1:34
```

I've also included the generated `cimport.zig` file for convenience:
```c
// .\bug.h:3:19: warning: TODO implement translation of CastKind BuiltinFnToFnPtr
pub const Func2 = @compileError ("unable to translate function")
pub const Func1 = @compileError("unable to translate function"); // .\bug.h:1:34
// .\bug.h:7:5: warning: TODO implement translation of CastKind BuiltinFnToFnPtr
pub const Func2 = @compileError("unable to translate function"); // .\bug.h:5:6
pub const __INTMAX_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):56:9
pub const __UINTMAX_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):60:9
pub const __PTRDIFF_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):67:9
pub const __INTPTR_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):71:9
pub const __SIZE_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):75:9
pub const __UINTPTR_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):90:9
pub const __INT64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):154:9
pub const __UINT64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):182:9
pub const __INT_LEAST64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):220:9
pub const __UINT_LEAST64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):224:9
pub const __INT_FAST64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):260:9
pub const __UINT_FAST64_TYPE__ = @compileError("unable to translate C expr: unexpected token .Keyword_long"); // (no file):264:9
pub const __llvm__ = 1;
pub const __clang__ = 1;
pub const __clang_major__ = 11;
pub const __clang_minor__ = 0;
pub const __clang_patchlevel__ = 0;
pub const __clang_version__ = "11.0.0 ";
pub const __ATOMIC_RELAXED = 0;
pub const __ATOMIC_CONSUME = 1;
pub const __ATOMIC_ACQUIRE = 2;
pub const __ATOMIC_RELEASE = 3;
pub const __ATOMIC_ACQ_REL = 4;
pub const __ATOMIC_SEQ_CST = 5;
pub const __OPENCL_MEMORY_SCOPE_WORK_ITEM = 0;
pub const __OPENCL_MEMORY_SCOPE_WORK_GROUP = 1;
pub const __OPENCL_MEMORY_SCOPE_DEVICE = 2;
pub const __OPENCL_MEMORY_SCOPE_ALL_SVM_DEVICES = 3;
pub const __OPENCL_MEMORY_SCOPE_SUB_GROUP = 4;
pub const __PRAGMA_REDEFINE_EXTNAME = 1;
pub const __VERSION__ = "Clang 11.0.0";
pub const __OBJC_BOOL_IS_BOOL = 0;
pub const __CONSTANT_CFSTRINGS__ = 1;
pub const __OPTIMIZE__ = 1;
pub const __ORDER_LITTLE_ENDIAN__ = 1234;
pub const __ORDER_BIG_ENDIAN__ = 4321;
pub const __ORDER_PDP_ENDIAN__ = 3412;
pub const __BYTE_ORDER__ = __ORDER_LITTLE_ENDIAN__;
pub const __LITTLE_ENDIAN__ = 1;
pub const __CHAR_BIT__ = 8;
pub const __SCHAR_MAX__ = 127;
pub const __SHRT_MAX__ = 32767;
pub const __INT_MAX__ = 2147483647;
pub const __LONG_MAX__ = @as(c_long, 2147483647);
pub const __LONG_LONG_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __WCHAR_MAX__ = 65535;
pub const __WINT_MAX__ = 65535;
pub const __INTMAX_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __SIZE_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __UINTMAX_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __PTRDIFF_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __INTPTR_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __UINTPTR_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __SIZEOF_DOUBLE__ = 8;
pub const __SIZEOF_FLOAT__ = 4;
pub const __SIZEOF_INT__ = 4;
pub const __SIZEOF_LONG__ = 4;
pub const __SIZEOF_LONG_DOUBLE__ = 8;
pub const __SIZEOF_LONG_LONG__ = 8;
pub const __SIZEOF_POINTER__ = 8;
pub const __SIZEOF_SHORT__ = 2;
pub const __SIZEOF_PTRDIFF_T__ = 8;
pub const __SIZEOF_SIZE_T__ = 8;
pub const __SIZEOF_WCHAR_T__ = 2;
pub const __SIZEOF_WINT_T__ = 2;
pub const __SIZEOF_INT128__ = 16;
pub const __INTMAX_FMTd__ = "lld";
pub const __INTMAX_FMTi__ = "lli";
pub const __INTMAX_C_SUFFIX__ = LL;
pub const __UINTMAX_FMTo__ = "llo";
pub const __UINTMAX_FMTu__ = "llu";
pub const __UINTMAX_FMTx__ = "llx";
pub const __UINTMAX_FMTX__ = "llX";
pub const __UINTMAX_C_SUFFIX__ = ULL;
pub const __INTMAX_WIDTH__ = 64;
pub const __PTRDIFF_FMTd__ = "lld";
pub const __PTRDIFF_FMTi__ = "lli";
pub const __PTRDIFF_WIDTH__ = 64;
pub const __INTPTR_FMTd__ = "lld";
pub const __INTPTR_FMTi__ = "lli";
pub const __INTPTR_WIDTH__ = 64;
pub const __SIZE_FMTo__ = "llo";
pub const __SIZE_FMTu__ = "llu";
pub const __SIZE_FMTx__ = "llx";
pub const __SIZE_FMTX__ = "llX";
pub const __SIZE_WIDTH__ = 64;
pub const __WCHAR_TYPE__ = c_ushort;
pub const __WCHAR_WIDTH__ = 16;
pub const __WINT_TYPE__ = c_ushort;
pub const __WINT_WIDTH__ = 16;
pub const __SIG_ATOMIC_WIDTH__ = 32;
pub const __SIG_ATOMIC_MAX__ = 2147483647;
pub const __CHAR16_TYPE__ = c_ushort;
pub const __CHAR32_TYPE__ = c_uint;
pub const __UINTMAX_WIDTH__ = 64;
pub const __UINTPTR_FMTo__ = "llo";
pub const __UINTPTR_FMTu__ = "llu";
pub const __UINTPTR_FMTx__ = "llx";
pub const __UINTPTR_FMTX__ = "llX";
pub const __UINTPTR_WIDTH__ = 64;
pub const __FLT_DENORM_MIN__ = @as(f32, 1.40129846e-45);
pub const __FLT_HAS_DENORM__ = 1;
pub const __FLT_DIG__ = 6;
pub const __FLT_DECIMAL_DIG__ = 9;
pub const __FLT_EPSILON__ = @as(f32, 1.19209290e-7);
pub const __FLT_HAS_INFINITY__ = 1;
pub const __FLT_HAS_QUIET_NAN__ = 1;
pub const __FLT_MANT_DIG__ = 24;
pub const __FLT_MAX_10_EXP__ = 38;
pub const __FLT_MAX_EXP__ = 128;
pub const __FLT_MAX__ = @as(f32, 3.40282347e+38);
pub const __FLT_MIN_10_EXP__ = -37;
pub const __FLT_MIN_EXP__ = -125;
pub const __FLT_MIN__ = @as(f32, 1.17549435e-38);
pub const __DBL_DENORM_MIN__ = 4.9406564584124654e-324;
pub const __DBL_HAS_DENORM__ = 1;
pub const __DBL_DIG__ = 15;
pub const __DBL_DECIMAL_DIG__ = 17;
pub const __DBL_EPSILON__ = 2.2204460492503131e-16;
pub const __DBL_HAS_INFINITY__ = 1;
pub const __DBL_HAS_QUIET_NAN__ = 1;
pub const __DBL_MANT_DIG__ = 53;
pub const __DBL_MAX_10_EXP__ = 308;
pub const __DBL_MAX_EXP__ = 1024;
pub const __DBL_MAX__ = 1.7976931348623157e+308;
pub const __DBL_MIN_10_EXP__ = -307;
pub const __DBL_MIN_EXP__ = -1021;
pub const __DBL_MIN__ = 2.2250738585072014e-308;
pub const __LDBL_DENORM_MIN__ = @as(c_longdouble, 4.9406564584124654e-324);
pub const __LDBL_HAS_DENORM__ = 1;
pub const __LDBL_DIG__ = 15;
pub const __LDBL_DECIMAL_DIG__ = 17;
pub const __LDBL_EPSILON__ = @as(c_longdouble, 2.2204460492503131e-16);
pub const __LDBL_HAS_INFINITY__ = 1;
pub const __LDBL_HAS_QUIET_NAN__ = 1;
pub const __LDBL_MANT_DIG__ = 53;
pub const __LDBL_MAX_10_EXP__ = 308;
pub const __LDBL_MAX_EXP__ = 1024;
pub const __LDBL_MAX__ = @as(c_longdouble, 1.7976931348623157e+308);
pub const __LDBL_MIN_10_EXP__ = -307;
pub const __LDBL_MIN_EXP__ = -1021;
pub const __LDBL_MIN__ = @as(c_longdouble, 2.2250738585072014e-308);
pub const __POINTER_WIDTH__ = 64;
pub const __BIGGEST_ALIGNMENT__ = 16;
pub const __WCHAR_UNSIGNED__ = 1;
pub const __WINT_UNSIGNED__ = 1;
pub const __INT8_TYPE__ = i8;
pub const __INT8_FMTd__ = "hhd";
pub const __INT8_FMTi__ = "hhi";
pub const __INT16_TYPE__ = c_short;
pub const __INT16_FMTd__ = "hd";
pub const __INT16_FMTi__ = "hi";
pub const __INT32_TYPE__ = c_int;
pub const __INT32_FMTd__ = "d";
pub const __INT32_FMTi__ = "i";
pub const __INT64_FMTd__ = "lld";
pub const __INT64_FMTi__ = "lli";
pub const __INT64_C_SUFFIX__ = LL;
pub const __UINT8_TYPE__ = u8;
pub const __UINT8_FMTo__ = "hho";
pub const __UINT8_FMTu__ = "hhu";
pub const __UINT8_FMTx__ = "hhx";
pub const __UINT8_FMTX__ = "hhX";
pub const __UINT8_MAX__ = 255;
pub const __INT8_MAX__ = 127;
pub const __UINT16_TYPE__ = c_ushort;
pub const __UINT16_FMTo__ = "ho";
pub const __UINT16_FMTu__ = "hu";
pub const __UINT16_FMTx__ = "hx";
pub const __UINT16_FMTX__ = "hX";
pub const __UINT16_MAX__ = 65535;
pub const __INT16_MAX__ = 32767;
pub const __UINT32_TYPE__ = c_uint;
pub const __UINT32_FMTo__ = "o";
pub const __UINT32_FMTu__ = "u";
pub const __UINT32_FMTx__ = "x";
pub const __UINT32_FMTX__ = "X";
pub const __UINT32_C_SUFFIX__ = U;
pub const __UINT32_MAX__ = @as(c_uint, 4294967295);
pub const __INT32_MAX__ = 2147483647;
pub const __UINT64_FMTo__ = "llo";
pub const __UINT64_FMTu__ = "llu";
pub const __UINT64_FMTx__ = "llx";
pub const __UINT64_FMTX__ = "llX";
pub const __UINT64_C_SUFFIX__ = ULL;
pub const __UINT64_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __INT64_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __INT_LEAST8_TYPE__ = i8;
pub const __INT_LEAST8_MAX__ = 127;
pub const __INT_LEAST8_FMTd__ = "hhd";
pub const __INT_LEAST8_FMTi__ = "hhi";
pub const __UINT_LEAST8_TYPE__ = u8;
pub const __UINT_LEAST8_MAX__ = 255;
pub const __UINT_LEAST8_FMTo__ = "hho";
pub const __UINT_LEAST8_FMTu__ = "hhu";
pub const __UINT_LEAST8_FMTx__ = "hhx";
pub const __UINT_LEAST8_FMTX__ = "hhX";
pub const __INT_LEAST16_TYPE__ = c_short;
pub const __INT_LEAST16_MAX__ = 32767;
pub const __INT_LEAST16_FMTd__ = "hd";
pub const __INT_LEAST16_FMTi__ = "hi";
pub const __UINT_LEAST16_TYPE__ = c_ushort;
pub const __UINT_LEAST16_MAX__ = 65535;
pub const __UINT_LEAST16_FMTo__ = "ho";
pub const __UINT_LEAST16_FMTu__ = "hu";
pub const __UINT_LEAST16_FMTx__ = "hx";
pub const __UINT_LEAST16_FMTX__ = "hX";
pub const __INT_LEAST32_TYPE__ = c_int;
pub const __INT_LEAST32_MAX__ = 2147483647;
pub const __INT_LEAST32_FMTd__ = "d";
pub const __INT_LEAST32_FMTi__ = "i";
pub const __UINT_LEAST32_TYPE__ = c_uint;
pub const __UINT_LEAST32_MAX__ = @as(c_uint, 4294967295);
pub const __UINT_LEAST32_FMTo__ = "o";
pub const __UINT_LEAST32_FMTu__ = "u";
pub const __UINT_LEAST32_FMTx__ = "x";
pub const __UINT_LEAST32_FMTX__ = "X";
pub const __INT_LEAST64_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __INT_LEAST64_FMTd__ = "lld";
pub const __INT_LEAST64_FMTi__ = "lli";
pub const __UINT_LEAST64_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __UINT_LEAST64_FMTo__ = "llo";
pub const __UINT_LEAST64_FMTu__ = "llu";
pub const __UINT_LEAST64_FMTx__ = "llx";
pub const __UINT_LEAST64_FMTX__ = "llX";
pub const __INT_FAST8_TYPE__ = i8;
pub const __INT_FAST8_MAX__ = 127;
pub const __INT_FAST8_FMTd__ = "hhd";
pub const __INT_FAST8_FMTi__ = "hhi";
pub const __UINT_FAST8_TYPE__ = u8;
pub const __UINT_FAST8_MAX__ = 255;
pub const __UINT_FAST8_FMTo__ = "hho";
pub const __UINT_FAST8_FMTu__ = "hhu";
pub const __UINT_FAST8_FMTx__ = "hhx";
pub const __UINT_FAST8_FMTX__ = "hhX";
pub const __INT_FAST16_TYPE__ = c_short;
pub const __INT_FAST16_MAX__ = 32767;
pub const __INT_FAST16_FMTd__ = "hd";
pub const __INT_FAST16_FMTi__ = "hi";
pub const __UINT_FAST16_TYPE__ = c_ushort;
pub const __UINT_FAST16_MAX__ = 65535;
pub const __UINT_FAST16_FMTo__ = "ho";
pub const __UINT_FAST16_FMTu__ = "hu";
pub const __UINT_FAST16_FMTx__ = "hx";
pub const __UINT_FAST16_FMTX__ = "hX";
pub const __INT_FAST32_TYPE__ = c_int;
pub const __INT_FAST32_MAX__ = 2147483647;
pub const __INT_FAST32_FMTd__ = "d";
pub const __INT_FAST32_FMTi__ = "i";
pub const __UINT_FAST32_TYPE__ = c_uint;
pub const __UINT_FAST32_MAX__ = @as(c_uint, 4294967295);
pub const __UINT_FAST32_FMTo__ = "o";
pub const __UINT_FAST32_FMTu__ = "u";
pub const __UINT_FAST32_FMTx__ = "x";
pub const __UINT_FAST32_FMTX__ = "X";
pub const __INT_FAST64_MAX__ = @as(c_longlong, 9223372036854775807);
pub const __INT_FAST64_FMTd__ = "lld";
pub const __INT_FAST64_FMTi__ = "lli";
pub const __UINT_FAST64_MAX__ = @as(c_ulonglong, 18446744073709551615);
pub const __UINT_FAST64_FMTo__ = "llo";
pub const __UINT_FAST64_FMTu__ = "llu";
pub const __UINT_FAST64_FMTx__ = "llx";
pub const __UINT_FAST64_FMTX__ = "llX";
pub const __FINITE_MATH_ONLY__ = 0;
pub const __CLANG_ATOMIC_BOOL_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_CHAR_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_CHAR16_T_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_CHAR32_T_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_WCHAR_T_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_SHORT_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_INT_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_LONG_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_LLONG_LOCK_FREE = 2;
pub const __CLANG_ATOMIC_POINTER_LOCK_FREE = 2;
pub const __PIC__ = 2;
pub const __pic__ = 2;
pub const __FLT_EVAL_METHOD__ = 0;
pub const __FLT_RADIX__ = 2;
pub const __DECIMAL_DIG__ = __LDBL_DECIMAL_DIG__;
pub const __SSP_STRONG__ = 2;
pub const __GCC_ASM_FLAG_OUTPUTS__ = 1;
pub const __code_model_small__ = 1;
pub const __amd64__ = 1;
pub const __amd64 = 1;
pub const __x86_64 = 1;
pub const __x86_64__ = 1;
pub const __SEG_GS = 1;
pub const __SEG_FS = 1;
pub const __seg_gs = __attribute__(address_space(256));
pub const __seg_fs = __attribute__(address_space(257));
pub const __corei7 = 1;
pub const __corei7__ = 1;
pub const __tune_corei7__ = 1;
pub const __NO_MATH_INLINES = 1;
pub const __AES__ = 1;
pub const __PCLMUL__ = 1;
pub const __LZCNT__ = 1;
pub const __RDRND__ = 1;
pub const __FSGSBASE__ = 1;
pub const __BMI__ = 1;
pub const __BMI2__ = 1;
pub const __POPCNT__ = 1;
pub const __RTM__ = 1;
pub const __MOVBE__ = 1;
pub const __FMA__ = 1;
pub const __F16C__ = 1;
pub const __FXSR__ = 1;
pub const __XSAVE__ = 1;
pub const __XSAVEOPT__ = 1;
pub const __INVPCID__ = 1;
pub const __AVX2__ = 1;
pub const __AVX__ = 1;
pub const __SSE4_2__ = 1;
pub const __SSE4_1__ = 1;
pub const __SSSE3__ = 1;
pub const __SSE3__ = 1;
pub const __SSE2__ = 1;
pub const __SSE2_MATH__ = 1;
pub const __SSE__ = 1;
pub const __SSE_MATH__ = 1;
pub const __MMX__ = 1;
pub const __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 = 1;
pub const __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 = 1;
pub const __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4 = 1;
pub const __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8 = 1;
pub const __GCC_HAVE_SYNC_COMPARE_AND_SWAP_16 = 1;
pub const _WIN32 = 1;
pub const _WIN64 = 1;
pub const _MSC_VER = 1925;
pub const _MSC_FULL_VER = 192528614;
pub const _MSC_BUILD = 1;
pub const _MSC_EXTENSIONS = 1;
pub const _INTEGRAL_MAX_BITS = 64;
pub const _M_X64 = 100;
pub const _M_AMD64 = 100;
pub const __STDC_HOSTED__ = 1;
pub const __STDC_VERSION__ = @as(c_long, 201710);
pub const __STDC_UTF_16__ = 1;
pub const __STDC_UTF_32__ = 1;
pub const _DEBUG = 1;
pub const SomeStruct = struct_SomeStruct;

```

--- a/lib/std/c.zig
+++ b/lib/std/c.zig
@@ -12,6 +12,7 @@ pub const Token = tokenizer.Token;
 pub const Tokenizer = tokenizer.Tokenizer;
 pub const parse = @import("c/parse.zig").parse;
 pub const ast = @import("c/ast.zig");
+pub const builtins = @import("c/builtins.zig");
 
 test "" {
     _ = tokenizer;

--- a/lib/std/c/builtins.zig
+++ b/lib/std/c/builtins.zig
@@ -1,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2020 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+
+const std = @import("std");
+
+pub inline fn __builtin_bswap16(val: u16) callconv(.C) u16 { return @byteSwap(u16, val); }
+pub inline fn __builtin_bswap32(val: u32) callconv(.C) u32 { return @byteSwap(u32, val); }
+pub inline fn __builtin_bswap64(val: u64) callconv(.C) u64 { return @byteSwap(u64, val); }
+
+pub inline fn __builtin_signbit(val: f64) callconv(.C) c_int { return @boolToInt(std.math.signbit(val)); }
+pub inline fn __builtin_signbitf(val: f32) callconv(.C) c_int { return @boolToInt(std.math.signbit(val)); }
+
+pub inline fn __builtin_popcount(val: c_uint) callconv(.C) c_int {
+    // popcount of a c_uint will never exceed the capacity of a c_int
+    @setRuntimeSafety(false);
+    return @bitCast(c_int, @as(c_uint, @popCount(c_uint, val)));
+}
+pub inline fn __builtin_ctz(val: c_uint) callconv(.C) c_int {
+    // Returns the number of trailing 0-bits in val, starting at the least significant bit position.
+    // In C if `val` is 0, the result is undefined; in zig it's the number of bits in a c_uint
+    @setRuntimeSafety(false);
+    return @bitCast(c_int, @as(c_uint, @ctz(c_uint, val)));
+}
+pub inline fn __builtin_clz(val: c_uint) callconv(.C) c_int {
+    // Returns the number of leading 0-bits in x, starting at the most significant bit position.
+    // In C if `val` is 0, the result is undefined; in zig it's the number of bits in a c_uint
+    @setRuntimeSafety(false);
+    return @bitCast(c_int, @as(c_uint, @clz(c_uint, val)));
+}
+
+pub inline fn __builtin_sqrt(val: f64) callconv(.C) f64 { return @sqrt(val); }
+pub inline fn __builtin_sqrtf(val: f32) callconv(.C) f32 { return @sqrt(val); }
+
+pub inline fn __builtin_sin(val: f64) callconv(.C) f64 { return @sin(val); }
+pub inline fn __builtin_sinf(val: f32) callconv(.C) f32 { return @sin(val); }
+pub inline fn __builtin_cos(val: f64) callconv(.C) f64 { return @cos(val); }
+pub inline fn __builtin_cosf(val: f32) callconv(.C) f32 { return @cos(val); }
+
+pub inline fn __builtin_exp(val: f64) callconv(.C) f64 { return @exp(val); }
+pub inline fn __builtin_expf(val: f32) callconv(.C) f32 { return @exp(val); }
+pub inline fn __builtin_exp2(val: f64) callconv(.C) f64 { return @exp2(val); }
+pub inline fn __builtin_exp2f(val: f32) callconv(.C) f32 { return @exp2(val); }
+pub inline fn __builtin_log(val: f64) callconv(.C) f64 { return @log(val); }
+pub inline fn __builtin_logf(val: f32) callconv(.C) f32 { return @log(val); }
+pub inline fn __builtin_log2(val: f64) callconv(.C) f64 { return @log2(val); }
+pub inline fn __builtin_log2f(val: f32) callconv(.C) f32 { return @log2(val); }
+pub inline fn __builtin_log10(val: f64) callconv(.C) f64 { return @log10(val); }
+pub inline fn __builtin_log10f(val: f32) callconv(.C) f32 { return @log10(val); }
+
+// Standard C Library bug: The absolute value of the most negative integer remains negative.
+pub inline fn __builtin_abs(val: c_int) callconv(.C) c_int { return std.math.absInt(val) catch std.math.minInt(c_int); }
+pub inline fn __builtin_fabs(val: f64) callconv(.C) f64 { return @fabs(val); }
+pub inline fn __builtin_fabsf(val: f32) callconv(.C) f32 { return @fabs(val); }
+
+pub inline fn __builtin_floor(val: f64) callconv(.C) f64 { return @floor(val); }
+pub inline fn __builtin_floorf(val: f32) callconv(.C) f32 { return @floor(val); }
+pub inline fn __builtin_ceil(val: f64) callconv(.C) f64 { return @ceil(val); }
+pub inline fn __builtin_ceilf(val: f32) callconv(.C) f32 { return @ceil(val); }
+pub inline fn __builtin_trunc(val: f64) callconv(.C) f64 { return @trunc(val); }
+pub inline fn __builtin_truncf(val: f32) callconv(.C) f32 { return @trunc(val); }
+pub inline fn __builtin_round(val: f64) callconv(.C) f64 { return @round(val); }
+pub inline fn __builtin_roundf(val: f32) callconv(.C) f32 { return @round(val); }
+
+pub inline fn __builtin_strlen(s: [*c]const u8) callconv(.C) usize { return std.mem.lenZ(s); }
+pub inline fn __builtin_strcmp(s1: [*c]const u8, s2: [*c]const u8) callconv(.C) c_int {
+    return @as(c_int, std.cstr.cmp(s1, s2));
+}
+
+pub inline fn __builtin_object_size(ptr: ?*const c_void, ty: c_int) callconv(.C) usize {
+    // clang semantics match gcc's: https://gcc.gnu.org/onlinedocs/gcc/Object-Size-Checking.html
+    // If it is not possible to determine which objects ptr points to at compile time,
+    // __builtin_object_size should return (size_t) -1 for type 0 or 1 and (size_t) 0
+    // for type 2 or 3.
+    if (ty == 0 or ty == 1) return @bitCast(usize, -@as(c_long, 1));
+    if (ty == 2 or ty == 3) return 0;
+    unreachable;
+}
+
+pub inline fn __builtin___memset_chk(
+    dst: ?*c_void,
+    val: c_int,
+    len: usize,
+    remaining: usize,
+) callconv(.C) ?*c_void {
+    if (len > remaining) @panic("std.c.builtins.memset_chk called with len > remaining");
+    return __builtin_memset(dst, val, len);
+}
+
+pub inline fn __builtin_memset(dst: ?*c_void, val: c_int, len: usize) callconv(.C) ?*c_void {
+    const dst_cast = @ptrCast([*c]u8, dst);
+    @memset(dst_cast, @bitCast(u8, @truncate(i8, val)), len);
+    return dst;
+}
+
+pub inline fn __builtin___memcpy_chk(
+    noalias dst: ?*c_void,
+    noalias src: ?*const c_void,
+    len: usize,
+    remaining: usize,
+) callconv(.C) ?*c_void {
+    if (len > remaining) @panic("std.c.builtins.memcpy_chk called with len > remaining");
+    return __builtin_memcpy(dst, src, len);
+}
+
+pub inline fn __builtin_memcpy(
+    noalias dst: ?*c_void,
+    noalias src: ?*const c_void,
+    len: usize,
+) callconv(.C) ?*c_void {
+    const dst_cast = @ptrCast([*c]u8, dst);
+    const src_cast = @ptrCast([*c]const u8, src);
+
+    @memcpy(dst_cast, src_cast, len);
+    return dst;
+}

--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -328,6 +328,47 @@ pub const Context = struct {
     }
 };
 
+fn addCBuiltinsNamespace(c: *Context) Error!void {
+    // pub usingnamespace @import("std").c.builtins;
+    const pub_tok = try appendToken(c, .Keyword_pub, "pub");
+    const use_tok = try appendToken(c, .Keyword_usingnamespace, "usingnamespace");
+    const import_tok = try appendToken(c, .Builtin, "@import");
+    const lparen_tok = try appendToken(c, .LParen, "(");
+    const std_tok = try appendToken(c, .StringLiteral, "\"std\"");
+    const rparen_tok = try appendToken(c, .RParen, ")");
+
+    const std_node = try c.arena.create(ast.Node.OneToken);
+    std_node.* = .{
+        .base = .{ .tag = .StringLiteral },
+        .token = std_tok,
+    };
+
+    const call_node = try ast.Node.BuiltinCall.alloc(c.arena, 1);
+    call_node.* = .{
+        .builtin_token = import_tok,
+        .params_len = 1,
+        .rparen_token = rparen_tok,
+    };
+    call_node.params()[0] = &std_node.base;
+
+    var access_chain = &call_node.base;
+    access_chain = try transCreateNodeFieldAccess(c, access_chain, "c");
+    access_chain = try transCreateNodeFieldAccess(c, access_chain, "builtins");
+
+    const semi_tok = try appendToken(c, .Semicolon, ";");
+
+    const bytes = try c.gpa.alignedAlloc(u8, @alignOf(ast.Node.Use), @sizeOf(ast.Node.Use));
+    const using_node = @ptrCast(*ast.Node.Use, bytes.ptr);
+    using_node.* = .{
+        .doc_comments = null,
+        .visib_token = pub_tok,
+        .use_token = use_tok,
+        .expr = access_chain,
+        .semicolon_token = semi_tok,
+    };
+    try c.root_decls.append(c.gpa, &using_node.base);
+}
+
 pub fn translate(
     gpa: *mem.Allocator,
     args_begin: [*]?[*]const u8,
@@ -376,6 +417,8 @@ pub fn translate(
         context.root_decls.deinit(gpa);
         context.opaque_demotes.deinit(gpa);
     }
+
+    try addCBuiltinsNamespace(&context);
 
     try prepopulateGlobalNameTable(ast_unit, &context);
 
@@ -1726,6 +1769,9 @@ fn transImplicitCastExpr(
             const rhs_node = try transCreateNodeInt(rp.c, 0);
             return transCreateNodeInfixOp(rp, scope, sub_expr_node, .BangEqual, op_token, rhs_node, result_used, false);
         },
+        .BuiltinFnToFnPtr => {
+            return transExpr(rp, scope, sub_expr, .used, .r_value);
+        },
         else => |kind| return revertAndWarn(
             rp,
             error.UnsupportedTranslation,
@@ -2028,7 +2074,6 @@ fn transCCast(
         // 1. If src_type is an enum, determine the underlying signed int type
         // 2. Extend or truncate without changing signed-ness.
         // 3. Bit-cast to correct signed-ness
-
         const src_type_is_signed = cIsSignedInteger(src_type) or cIsEnum(src_type);
         const src_int_type = if (cIsInteger(src_type)) src_type else cIntTypeForEnum(src_type);
         const src_int_expr = if (cIsInteger(src_type)) expr else try transEnumToInt(rp.c, expr);
@@ -3048,8 +3093,9 @@ fn transCallExpr(rp: RestorePoint, scope: *Scope, stmt: *const clang.CallExpr, r
     const fn_expr = if (is_ptr and fn_ty != null) blk: {
         if (callee.getStmtClass() == .ImplicitCastExprClass) {
             const implicit_cast = @ptrCast(*const clang.ImplicitCastExpr, callee);
-
-            if (implicit_cast.getCastKind() == .FunctionToPointerDecay) {
+            const cast_kind = implicit_cast.getCastKind();
+            if (cast_kind == .BuiltinFnToFnPtr) break :blk raw_fn_expr;
+            if (cast_kind == .FunctionToPointerDecay) {
                 const subexpr = implicit_cast.getSubExpr();
                 if (subexpr.getStmtClass() == .DeclRefExprClass) {
                     const decl_ref = @ptrCast(*const clang.DeclRefExpr, subexpr);
@@ -5881,7 +5927,7 @@ fn parseCPrimaryExprInner(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!*
                 return error.ParseError;
             }
 
-            const ident_token = try appendTokenFmt(c, .Identifier, "{}_{}", .{slice, m.slice()});
+            const ident_token = try appendTokenFmt(c, .Identifier, "{}_{}", .{ slice, m.slice() });
             const identifier = try c.arena.create(ast.Node.OneToken);
             identifier.* = .{
                 .base = .{ .tag = .Identifier },

--- a/test/run_translated_c.zig
+++ b/test/run_translated_c.zig
@@ -491,4 +491,157 @@ pub fn addCases(cases: *tests.RunTranslatedCContext) void {
         \\    return 0;
         \\}
     , "");
+
+    cases.add("issue #6707 cast builtin call result to opaque struct pointer",
+        \\#include <stdlib.h>
+        \\struct foo* make_foo(void)
+        \\{
+        \\    return (struct foo*)__builtin_strlen("0123456789ABCDEF");
+        \\}
+        \\int main(void) {
+        \\    struct foo *foo_pointer = make_foo();
+        \\    if (foo_pointer != (struct foo*)16) abort();
+        \\    return 0;
+        \\}
+    , "");
+
+    cases.add("C built-ins",
+        \\#include <stdlib.h>
+        \\#include <limits.h>
+        \\#include <stdbool.h>
+        \\#define M_E    2.71828182845904523536
+        \\#define M_PI_2 1.57079632679489661923
+        \\bool check_clz(unsigned int pos) {
+        \\    return (__builtin_clz(1 << pos) == (8 * sizeof(unsigned int) - pos - 1));
+        \\}
+        \\int main(void) {
+        \\    if (__builtin_bswap16(0x0102) != 0x0201) abort();
+        \\    if (__builtin_bswap32(0x01020304) != 0x04030201) abort();
+        \\    if (__builtin_bswap64(0x0102030405060708) != 0x0807060504030201) abort();
+        \\
+        \\    if (__builtin_signbit(0.0) != 0) abort();
+        \\    if (__builtin_signbitf(0.0f) != 0) abort();
+        \\    if (__builtin_signbit(1.0) != 0) abort();
+        \\    if (__builtin_signbitf(1.0f) != 0) abort();
+        \\    if (__builtin_signbit(-1.0) != 1) abort();
+        \\    if (__builtin_signbitf(-1.0f) != 1) abort();
+        \\
+        \\    if (__builtin_popcount(0) != 0) abort();
+        \\    if (__builtin_popcount(0b1) != 1) abort();
+        \\    if (__builtin_popcount(0b11) != 2) abort();
+        \\    if (__builtin_popcount(0b1111) != 4) abort();
+        \\    if (__builtin_popcount(0b11111111) != 8) abort();
+        \\
+        \\    if (__builtin_ctz(0b1) != 0) abort();
+        \\    if (__builtin_ctz(0b10) != 1) abort();
+        \\    if (__builtin_ctz(0b100) != 2) abort();
+        \\    if (__builtin_ctz(0b10000) != 4) abort();
+        \\    if (__builtin_ctz(0b100000000) != 8) abort();
+        \\
+        \\    if (!check_clz(0)) abort();
+        \\    if (!check_clz(1)) abort();
+        \\    if (!check_clz(2)) abort();
+        \\    if (!check_clz(4)) abort();
+        \\    if (!check_clz(8)) abort();
+        \\
+        \\    if (__builtin_sqrt(__builtin_sqrt(__builtin_sqrt(256))) != 2.0) abort();
+        \\    if (__builtin_sqrt(__builtin_sqrt(__builtin_sqrt(256.0))) != 2.0) abort();
+        \\    if (__builtin_sqrt(__builtin_sqrt(__builtin_sqrt(256.0f))) != 2.0) abort();
+        \\    if (__builtin_sqrtf(__builtin_sqrtf(__builtin_sqrtf(256.0f))) != 2.0f) abort();
+        \\
+        \\    if (__builtin_sin(1.0) != -__builtin_sin(-1.0)) abort();
+        \\    if (__builtin_sinf(1.0f) != -__builtin_sinf(-1.0f)) abort();
+        \\    if (__builtin_sin(M_PI_2) != 1.0) abort();
+        \\    if (__builtin_sinf(M_PI_2) != 1.0f) abort();
+        \\
+        \\    if (__builtin_cos(1.0) != __builtin_cos(-1.0)) abort();
+        \\    if (__builtin_cosf(1.0f) != __builtin_cosf(-1.0f)) abort();
+        \\    if (__builtin_cos(0.0) != 1.0) abort();
+        \\    if (__builtin_cosf(0.0f) != 1.0f) abort();
+        \\
+        \\    if (__builtin_exp(0) != 1.0) abort();
+        \\    if (__builtin_fabs(__builtin_exp(1.0) - M_E) > 0.00000001) abort();
+        \\    if (__builtin_exp(0.0f) != 1.0f) abort();
+        \\
+        \\    if (__builtin_exp2(0) != 1.0) abort();
+        \\    if (__builtin_exp2(4.0) != 16.0) abort();
+        \\    if (__builtin_exp2f(0.0f) != 1.0f) abort();
+        \\    if (__builtin_exp2f(4.0f) != 16.0f) abort();
+        \\
+        \\    if (__builtin_log(M_E) != 1.0) abort();
+        \\    if (__builtin_log(1.0) != 0.0) abort();
+        \\    if (__builtin_logf(1.0f) != 0.0f) abort();
+        \\
+        \\    if (__builtin_log2(8.0) != 3.0) abort();
+        \\    if (__builtin_log2(1.0) != 0.0) abort();
+        \\    if (__builtin_log2f(8.0f) != 3.0f) abort();
+        \\    if (__builtin_log2f(1.0f) != 0.0f) abort();
+        \\
+        \\    if (__builtin_log10(1000.0) != 3.0) abort();
+        \\    if (__builtin_log10(1.0) != 0.0) abort();
+        \\    if (__builtin_log10f(1000.0f) != 3.0f) abort();
+        \\    if (__builtin_log10f(1.0f) != 0.0f) abort();
+        \\
+        \\    if (__builtin_fabs(-42.0f) != 42.0) abort();
+        \\    if (__builtin_fabs(-42.0) != 42.0) abort();
+        \\    if (__builtin_fabs(-42) != 42.0) abort();
+        \\    if (__builtin_fabsf(-42.0f) != 42.0f) abort();
+        \\
+        \\    if (__builtin_fabs(-42.0f) != 42.0) abort();
+        \\    if (__builtin_fabs(-42.0) != 42.0) abort();
+        \\    if (__builtin_fabs(-42) != 42.0) abort();
+        \\    if (__builtin_fabsf(-42.0f) != 42.0f) abort();
+        \\
+        \\    if (__builtin_abs(42) != 42) abort();
+        \\    if (__builtin_abs(-42) != 42) abort();
+        \\    if (__builtin_abs(INT_MIN) != INT_MIN) abort();
+        \\
+        \\    if (__builtin_floor(42.9) != 42.0) abort();
+        \\    if (__builtin_floor(-42.9) != -43.0) abort();
+        \\    if (__builtin_floorf(42.9f) != 42.0f) abort();
+        \\    if (__builtin_floorf(-42.9f) != -43.0f) abort();
+        \\
+        \\    if (__builtin_ceil(42.9) != 43.0) abort();
+        \\    if (__builtin_ceil(-42.9) != -42) abort();
+        \\    if (__builtin_ceilf(42.9f) != 43.0f) abort();
+        \\    if (__builtin_ceilf(-42.9f) != -42.0f) abort();
+        \\
+        \\    if (__builtin_trunc(42.9) != 42.0) abort();
+        \\    if (__builtin_truncf(42.9f) != 42.0f) abort();
+        \\    if (__builtin_trunc(-42.9) != -42.0) abort();
+        \\    if (__builtin_truncf(-42.9f) != -42.0f) abort();
+        \\
+        \\    if (__builtin_round(0.5) != 1.0) abort();
+        \\    if (__builtin_round(-0.5) != -1.0) abort();
+        \\    if (__builtin_roundf(0.5f) != 1.0f) abort();
+        \\    if (__builtin_roundf(-0.5f) != -1.0f) abort();
+        \\
+        \\    if (__builtin_strcmp("abc", "abc") != 0) abort();
+        \\    if (__builtin_strcmp("abc", "def") >= 0 ) abort();
+        \\    if (__builtin_strcmp("def", "abc") <= 0) abort();
+        \\
+        \\    if (__builtin_strlen("this is a string") != 16) abort();
+        \\
+        \\    char *s = malloc(6);
+        \\    __builtin_memcpy(s, "hello", 5);
+        \\    s[5] = '\0';
+        \\    if (__builtin_strlen(s) != 5) abort();
+        \\
+        \\    __builtin_memset(s, 42, __builtin_strlen(s));
+        \\    if (s[0] != 42 || s[1] != 42 || s[2] != 42 || s[3] != 42 || s[4] != 42) abort();
+        \\
+        \\    free(s);
+        \\
+        \\    return 0;
+        \\}
+    , "");
+
+    cases.add("function macro that uses builtin",
+        \\#include <stdlib.h>
+        \\#define FOO(x, y) (__builtin_popcount((x)) + __builtin_strlen((y)))
+        \\int main() {
+        \\    if (FOO(7, "hello!") != 9) abort();
+        \\    return 0;
+        \\}
+    , "");
 }

